File: unity/Assets/Scripts/Mapping/WorldGenerator.cs
using System;
using System.Collections.Generic;

namespace GameDev.Mapping
{
    public enum WorldZone
    {
        CBD = 0,
        Midtown = 1,
        Residential = 2
    }

    [Serializable]
    public sealed class WorldConfig
    {
        public uint worldSeed = 1;
        public int chunkCols = 8;
        public int chunkRows = 6;
        public int chunkWidth = 256;
        public int chunkHeight = 256;
    }

    [Serializable]
    public sealed class ChunkPlan
    {
        public int cx;
        public int cy;
        public uint chunkSeed;
        public string cityProfile = "";
        public CoastSide coastSide = CoastSide.None;
        public WorldZone worldZone = WorldZone.Residential;
    }

    public sealed class WorldGenerator
    {
        private const uint SaltWorldHighwayNs = 0xDEADBEEFu;
        private const uint SaltWorldHighwayEw = 0xBEEFCAFEu;

        private readonly WorldConfig config;
        private readonly List<ChunkPlan> chunks = new List<ChunkPlan>();

        public WorldGenerator(WorldConfig config)
        {
            if (config == null)
            {
                throw new ArgumentNullException("config");
            }

            if (config.chunkCols <= 0 || config.chunkRows <= 0)
            {
                throw new ArgumentException("WorldConfig chunk dimensions must be positive");
            }

            if (config.chunkWidth <= 0 || config.chunkHeight <= 0)
            {
                throw new ArgumentException("WorldConfig chunk size must be positive");
            }

            this.config = config;
        }

        public IList<ChunkPlan> Chunks
        {
            get { return chunks.AsReadOnly(); }
        }

        public void Plan()
        {
            chunks.Clear();
            chunks.Capacity = Math.Max(chunks.Capacity, config.chunkCols * config.chunkRows);

            for (int cy = 0; cy < config.chunkRows; ++cy)
            {
                for (int cx = 0; cx < config.chunkCols; ++cx)
                {
                    WorldZone zone = WorldZoneForChunk(cx, cy, config.chunkCols, config.chunkRows);
                    chunks.Add(new ChunkPlan
                    {
                        cx = cx,
                        cy = cy,
                        chunkSeed = DeriveChunkSeed(config.worldSeed, cx, cy),
                        cityProfile = ProfileForChunk(cx, cy, config.chunkCols, config.chunkRows),
                        coastSide = CoastForChunk(cx, cy, config.chunkCols, config.chunkRows),
                        worldZone = zone
                    });
                }
            }
        }

        public int ChunkCount()
        {
            EnsurePlanned();
            return chunks.Count;
        }

        public ChunkPlan ChunkAt(int cx, int cy)
        {
            EnsurePlanned();
            if (cx < 0 || cx >= config.chunkCols || cy < 0 || cy >= config.chunkRows)
            {
                throw new ArgumentOutOfRangeException("cx", "Chunk coordinates out of range");
            }

            return chunks[cy * config.chunkCols + cx];
        }

        public MapConfig ConfigForChunk(int cx, int cy)
        {
            ChunkPlan plan = ChunkAt(cx, cy);
            MapConfig mapConfig = new MapConfig();
            mapConfig.masterSeed = plan.chunkSeed;
            mapConfig.cityProfile = plan.cityProfile;
            mapConfig.coastSide = plan.coastSide;
            mapConfig.width = config.chunkWidth;
            mapConfig.height = config.chunkHeight;
            mapConfig.enableLocalRivers = false;
            mapConfig.hasWorldChunkMetadata = true;
            mapConfig.worldSeed = config.worldSeed;
            mapConfig.chunkX = cx;
            mapConfig.chunkY = cy;
            mapConfig.chunkSize = config.chunkWidth;
            ApplyHighwayBoundaryConstraints(mapConfig, cx, cy);
            return mapConfig;
        }

        public List<int> HighwayNsColumns()
        {
            return HighwayNsColumns(config.worldSeed, config.chunkCols);
        }

        public List<int> HighwayEwRows()
        {
            return HighwayEwRows(config.worldSeed, config.chunkRows);
        }

        public static uint DeriveChunkSeed(uint worldSeed, int cx, int cy)
        {
            unchecked
            {
                uint mixed = worldSeed ^
                    ((uint)cx * 0x9e3779b9u) ^
                    ((uint)cy * 0x517cc1b7u);
                return MapGenerator.MixU32(mixed);
            }
        }

        public static WorldZone WorldZoneForChunk(int cx, int cy, int chunkCols = 8, int chunkRows = 6)
        {
            double nx = Math.Abs(cx - (chunkCols - 1) / 2.0) / (chunkCols / 2.0);
            double ny = Math.Abs(cy - (chunkRows - 1) / 2.0) / (chunkRows / 2.0);
            double dist = Math.Max(nx, ny);

            if (dist <= 0.30) return WorldZone.CBD;
            if (dist <= 0.65) return WorldZone.Midtown;
            return WorldZone.Residential;
        }

        public static string ProfileForChunk(int cx, int cy, int chunkCols = 8, int chunkRows = 6)
        {
            WorldZone zone = WorldZoneForChunk(cx, cy, chunkCols, chunkRows);
            switch (zone)
            {
                case WorldZone.CBD:
                    return (cx + cy) % 2 == 0 ? "manhattan" : "paris_haussmann";
                case WorldZone.Midtown:
                    return "generic_dense";
                default:
                    return "london_organic";
            }
        }

        public static CoastSide CoastForChunk(int cx, int cy, int chunkCols = 8, int chunkRows = 6)
        {
            if (cy == 0) return CoastSide.North;
            if (cy == chunkRows - 1) return CoastSide.South;
            if (cx == 0) return CoastSide.West;
            if (cx == chunkCols - 1) return CoastSide.East;
            return CoastSide.None;
        }

        public static List<int> HighwayNsColumns(uint worldSeed, int chunkCols = 8)
        {
            uint h = MapGenerator.MixU32(worldSeed ^ SaltWorldHighwayNs);
            int offset = (int)(h % 2u);
            List<int> columns = new List<int> { 1 + offset, 3, 5 + offset };
            return FilterInterior(columns, chunkCols);
        }

        public static List<int> HighwayEwRows(uint worldSeed, int chunkRows = 6)
        {
            uint h = MapGenerator.MixU32(worldSeed ^ SaltWorldHighwayEw);
            int offset = (int)(h % 2u);
            List<int> rows = new List<int> { 1 + offset, 4 };
            return FilterInterior(rows, chunkRows);
        }

        private static List<int> FilterInterior(List<int> values, int upperBound)
        {
            List<int> filtered = new List<int>();
            for (int i = 0; i < values.Count; ++i)
            {
                int value = values[i];
                if (value > 0 && value < upperBound - 1 && !filtered.Contains(value))
                {
                    filtered.Add(value);
                }
            }

            filtered.Sort();
            return filtered;
        }

        private void ApplyHighwayBoundaryConstraints(MapConfig mapConfig, int cx, int cy)
        {
            List<int> nsColumns = HighwayNsColumns();
            List<int> ewRows = HighwayEwRows();

            mapConfig.forceWestHighway = nsColumns.Contains(cx - 1);
            mapConfig.forceEastHighway = nsColumns.Contains(cx);
            mapConfig.forceNorthHighway = ewRows.Contains(cy - 1);
            mapConfig.forceSouthHighway = ewRows.Contains(cy);
        }

        private void EnsurePlanned()
        {
            if (chunks.Count == 0)
            {
                Plan();
            }
        }
    }
}

File: mapping_algorithm/cpp/src/world_generator.cpp
#include "mapping_algorithm/world_generator.hpp"
#include <stdexcept>

namespace mapping_algorithm {

namespace {

std::uint32_t mix_u32_wg(std::uint32_t x) {
    x ^= x >> 16;
    x *= 0x7feb352dU;
    x ^= x >> 15;
    x *= 0x846ca68bU;
    x ^= x >> 16;
    return x;
}

} // namespace

WorldGenerator::WorldGenerator(WorldConfig config)
    : world_config_(std::move(config)) {
    if (world_config_.chunk_cols <= 0 || world_config_.chunk_rows <= 0)
        throw std::invalid_argument("WorldConfig chunk dimensions must be positive");
    if (world_config_.chunk_width <= 0 || world_config_.chunk_height <= 0)
        throw std::invalid_argument("WorldConfig chunk size must be positive");
}

void WorldGenerator::plan() {
    chunks_.clear();
    chunks_.reserve(static_cast<std::size_t>(world_config_.chunk_cols * world_config_.chunk_rows));
    for (int cy = 0; cy < world_config_.chunk_rows; ++cy) {
        for (int cx = 0; cx < world_config_.chunk_cols; ++cx) {
            ChunkPlan plan;
            plan.cx = cx;
            plan.cy = cy;
            plan.chunk_seed = derive_chunk_seed(world_config_.world_seed, cx, cy);
            plan.city_profile = profile_for_chunk(cx, cy, world_config_.chunk_cols, world_config_.chunk_rows);
            plan.coast_side = coast_for_chunk(cx, cy, world_config_.chunk_cols, world_config_.chunk_rows);
            chunks_.push_back(plan);
        }
    }
}

int WorldGenerator::chunk_count() const noexcept {
    return static_cast<int>(chunks_.size());
}

const ChunkPlan& WorldGenerator::chunk_at(int cx, int cy) const {
    if (cx < 0 || cx >= world_config_.chunk_cols || cy < 0 || cy >= world_config_.chunk_rows)
        throw std::out_of_range("Chunk coordinates out of range");
    return chunks_[static_cast<std::size_t>(cy * world_config_.chunk_cols + cx)];
}

MapConfig WorldGenerator::config_for_chunk(int cx, int cy) const {
    const auto& plan = chunk_at(cx, cy);
    MapConfig cfg;
    cfg.master_seed = plan.chunk_seed;
    cfg.city_profile = plan.city_profile;
    cfg.coast_side = plan.coast_side;
    cfg.width = world_config_.chunk_width;
    cfg.height = world_config_.chunk_height;
    return cfg;
}

std::uint32_t WorldGenerator::derive_chunk_seed(std::uint32_t world_seed, int cx, int cy) noexcept {
    // Chunk seed formula from OPEN_WORLD_SPEC.md:
    // mix_u32(world_seed XOR cx*0x9e3779b9 XOR cy*0x517cc1b7)
    const std::uint32_t mixed =
        world_seed ^
        (static_cast<std::uint32_t>(cx) * 0x9e3779b9U) ^
        (static_cast<std::uint32_t>(cy) * 0x517cc1b7U);
    return mix_u32_wg(mixed);
}

std::string WorldGenerator::profile_for_chunk(int cx, int cy, int chunk_cols, int chunk_rows) noexcept {
    // World zone grid from OPEN_WORLD_SPEC.md:
    // CBD centre: cx 2–5, cy 1–4 → alternate manhattan / paris_haussmann
    // Midtown ring (one chunk out from CBD) → generic_dense
    // Residential outer → london_organic or barcelona_eixample
    const int cx_cbd_lo = chunk_cols / 4;
    const int cx_cbd_hi = (chunk_cols * 3) / 4 - 1;
    const int cy_cbd_lo = chunk_rows / 5;
    const int cy_cbd_hi = (chunk_rows * 4) / 5 - 1;

    const bool in_cbd = cx >= cx_cbd_lo && cx <= cx_cbd_hi && cy >= cy_cbd_lo && cy <= cy_cbd_hi;
    if (in_cbd) {
        return ((cx + cy) % 2 == 0) ? "manhattan" : "paris_haussmann";
    }

    const int cx_mid_lo = cx_cbd_lo - 1;
    const int cx_mid_hi = cx_cbd_hi + 1;
    const int cy_mid_lo = cy_cbd_lo - 1;
    const int cy_mid_hi = cy_cbd_hi + 1;
    const bool in_midtown = cx >= cx_mid_lo && cx <= cx_mid_hi && cy >= cy_mid_lo && cy <= cy_mid_hi;
    if (in_midtown) return "generic_dense";

    return (cx % 2 == 0) ? "london_organic" : "barcelona_eixample";
}

CoastSide WorldGenerator::coast_for_chunk(int cx, int cy, int chunk_cols, int chunk_rows) noexcept {
    // World edges get coastal chunks per OPEN_WORLD_SPEC.md
    if (cy == 0) return CoastSide::North;
    if (cy == chunk_rows - 1) return CoastSide::South;
    if (cx == 0) return CoastSide::West;
    if (cx == chunk_cols - 1) return CoastSide::East;
    return CoastSide::None;
}

} // namespace mapping_algorithm
